Sources/_MatchingEngine/Regex/AST/AST.swift

```diff
@@ -207,14 +210,61 @@ extension AST {
 
 // FIXME: Get this out of here
 public struct CaptureTransform: Equatable, Hashable, CustomStringConvertible {
-  public let closure: (Substring) -> Any
+  public enum Closure {
+    case nonfailable((Substring) -> Any)
+    case failable((Substring) -> Any?)
+    case throwing((Substring) throws -> Any)
+  }
+  public let resultType: Any.Type
+  public let closure: Closure
 
-  public init(_ closure: @escaping (Substring) -> Any) {
+  public init(resultType: Any.Type, closure: Closure) {
+    self.resultType = resultType
     self.closure = closure
   }
 
-  public func callAsFunction(_ input: Substring) -> Any {
-    closure(input)
+  public init(
+    resultType: Any.Type,
+    _ closure: @escaping (Substring) -> Any
+  ) {
+    self.init(resultType: resultType, closure: .nonfailable(closure))
+  }
+
+  public init(
+    resultType: Any.Type,
+    _ closure: @escaping (Substring) -> Any?
+  ) {
+    self.init(resultType: resultType, closure: .failable(closure))
+  }
+
+  public init(
+    resultType: Any.Type,
+    _ closure: @escaping (Substring) throws -> Any
+  ) {
+    self.init(resultType: resultType, closure: .throwing(closure))
+  }
+
+  public func callAsFunction(_ input: Substring) -> Any? {
```

Comment:
Should this be `throws` then?

Reply:
No, this returns an optional for the engine to determine whether to fail matching. When the closure throws, this returns nil.

Reply:
> Should we have a throw abort?
Just saw this comment. In that case I think we can have if.

Reply:
SGTM